aom/src/main/java/com/nedap/archie/query/AOMPathQuery.java

```diff
@@ -225,18 +226,47 @@ protected ArchetypeModelObject findOneMatchingObject(CAttribute attribute, PathS
      * Useful mainly when flattening, probably does not have many other uses
      */
     public CComplexObjectProxy findAnyInternalReference(CComplexObject root) {
+       return (CComplexObjectProxy) findMatchingPredicate(root, (o) -> o instanceof CComplexObjectProxy);
+    }
+
+    /**
+     * Find anything matching a specific predicate anywhere inside the APath query. Can be at the end of the full query, at the first matching CComplexObjectProxy or anywhere in between
+     * Useful mainly when flattening, probably does not have many other uses
+     */
+    public ArchetypeModelObject findMatchingPredicate(CComplexObject root, Predicate predicate) {
         List<ArchetypeModelObject> result = new ArrayList<>();
         result.add(root);
         for(PathSegment segment:this.pathSegments) {
             if (result.size() == 0) {
                 return null;
             }
             result = findOneSegment(segment, result, false);
-            if(result.size() == 1 && result.get(0) instanceof CComplexObjectProxy) {
-                return (CComplexObjectProxy) result.get(0);
+            if(result.size() == 1 && predicate.test(result.get(0))) {
+                return result.get(0);
             }
         }
         return null;
 
     }
+
+    /**
+     * Find all path segments matching a specific predicate anywhere inside the APath query. Can be at the end of the full query, at the first matching CComplexObjectProxy or anywhere in between
+     * Useful mainly when flattening, probably does not have many other uses
+     */
+    public List<ArchetypeModelObject> findAllMatchingPredicate(CComplexObject root, Predicate predicate) {
+        List<ArchetypeModelObject> result = new ArrayList<>();
+        result.add(root);
+        List<ArchetypeModelObject> results = new ArrayList<>();
```

Comment:
Having two variables named `result` and `results` is a bit confusing. Maybe rename `result` to something like `currentObjects`?

Reply:
oh wow that is ugly indeed. I'll rewrite this.